x-pack/plugins/canvas/server/routes/socket.js

```diff
@@ -12,22 +12,33 @@ import { functionsRegistry } from '../../common/lib/functions_registry';
 import { typesRegistry } from '../../common/lib/types_registry';
 import { loadServerPlugins } from '../lib/load_server_plugins';
 import { getRequest } from '../lib/get_request';
+import { API_ROUTE } from '../../common/lib/constants';
 
 export function socketApi(server) {
+  // add a POST ping route for `getRequest` to use
+  // TODO: remove this once we have upstream socket support
+  server.route({
+    method: 'POST',
+    path: `${API_ROUTE}/ping`,
+    handler: () => 'pong',
+  });
+
   const io = socket(server.listener, { path: '/socket.io' });
 
-  io.on('connection', socket => {
-    // Create the function list
-    socket.emit('getFunctionList');
-    const getClientFunctions = new Promise(resolve => socket.once('functionList', resolve));
+  io.on('connection', async socket => {
+    try {
+      const request = await getRequest(server, socket.handshake);
 
-    socket.on('getFunctionList', () => {
-      loadServerPlugins().then(() => socket.emit('functionList', functionsRegistry.toJS()));
-    });
+      // Create the function list
+      socket.emit('getFunctionList');
+      const getClientFunctions = new Promise(resolve => socket.once('functionList', resolve));
 
-    const handler = ({ ast, context, id }) => {
-      Promise.all([getClientFunctions, getRequest(server, socket.handshake)]).then(
-        ([clientFunctions, request]) => {
+      socket.on('getFunctionList', () => {
+        loadServerPlugins().then(() => socket.emit('functionList', functionsRegistry.toJS()));
+      });
+
+      const handler = ({ ast, context, id }) => {
+        Promise.all([getClientFunctions]).then(([clientFunctions]) => {
```

Comment:
This is interesting... for my own info: is this just a shortcut to execute the Promise you defined above, or is there a fun feature I'm not aware of here?

Reply:
Yeah, it just ensures that we wait for the promise to resolve before we execute any of the code in the handler function, since it'll fail before the client functions are loaded. This doesn't need to be wrapped in a `Promise.all` anymore though.